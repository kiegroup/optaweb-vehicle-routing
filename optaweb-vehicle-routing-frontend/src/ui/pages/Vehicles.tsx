/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Button,
  DataList,
  GutterSize,
  Split,
  SplitItem,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import * as React from 'react';
import { connect } from 'react-redux';
import { routeOperations } from 'store/route';
import { Vehicle } from 'store/route/types';
import { AppState } from 'store/types';
import VehicleItem from 'ui/components/Vehicle';

interface StateProps {
  vehicles: Vehicle[];
}

interface DispatchProps {
  addVehicleHandler: typeof routeOperations.addVehicle;
  removeVehicleHandler: typeof routeOperations.deleteVehicle;
  changeVehicleCapacityHandler: typeof routeOperations.changeVehicleCapacity;
}

export type Props = StateProps & DispatchProps;

const mapStateToProps = ({ plan }: AppState): StateProps => ({
  vehicles: plan.vehicles,
});

const mapDispatchToProps: DispatchProps = {
  addVehicleHandler: routeOperations.addVehicle,
  removeVehicleHandler: routeOperations.deleteVehicle,
  changeVehicleCapacityHandler: routeOperations.changeVehicleCapacity,
};

export const Vehicles: React.FC<Props> = ({
  vehicles, addVehicleHandler, removeVehicleHandler, changeVehicleCapacityHandler,
}) => (
  <>
    <Split gutter={GutterSize.md} style={{ overflowY: 'auto' }}>
      <SplitItem isFilled>
        <TextContent>
          <Text component={TextVariants.h1}>{`Vehicles (${vehicles.length})`}</Text>
        </TextContent>
      </SplitItem>
      <SplitItem isFilled={false}>
        <Button
          style={{ marginBottom: 16, marginLeft: 16 }}
          onClick={addVehicleHandler}
        >
          Add
        </Button>
      </SplitItem>
    </Split>
    <div style={{ overflowY: 'auto' }}>
      <DataList
        aria-label="List of vehicles"
      >
        {vehicles
          .slice(0) // clone the array because
          // sort is done in place (that would affect the route)
          .sort((a, b) => a.id - b.id)
          .map(vehicle => (
            <VehicleItem
              key={vehicle.id}
              id={vehicle.id}
              description={vehicle.name}
              capacity={vehicle.capacity}
              removeHandler={removeVehicleHandler}
              capacityChangeHandler={changeVehicleCapacityHandler}
            />
          ))}
      </DataList>
    </div>
  </>
);

export default connect(mapStateToProps, mapDispatchToProps)(Vehicles);
