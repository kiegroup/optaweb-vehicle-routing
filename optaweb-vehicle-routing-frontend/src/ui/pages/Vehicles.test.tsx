/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Props, Vehicles } from './Vehicles';

describe('Vehicles page', () => {
  it('should render correctly', () => {
    const props: Props = {
      addVehicleHandler: jest.fn(),
      removeVehicleHandler: jest.fn(),
      changeVehicleCapacityHandler: jest.fn,
      vehicles: [
        { id: 1, name: 'Vehicle 1', capacity: 5 },
        { id: 2, name: 'Vehicle 2', capacity: 5 },
      ],
    };
    const vehicles = shallow(<Vehicles {...props} />);
    expect(toJson(vehicles)).toMatchSnapshot();
  });
});
