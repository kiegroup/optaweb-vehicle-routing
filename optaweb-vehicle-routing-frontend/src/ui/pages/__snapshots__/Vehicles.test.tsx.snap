// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vehicles page should render correctly 1`] = `
<Fragment>
  <Split
    gutter="md"
    style={
      Object {
        "overflowY": "auto",
      }
    }
  >
    <SplitItem
      isFilled={true}
    >
      <TextContent>
        <Text
          component="h1"
        >
          Vehicles (2)
        </Text>
      </TextContent>
    </SplitItem>
    <SplitItem
      isFilled={false}
    >
      <Component
        onClick={[MockFunction]}
        style={
          Object {
            "marginBottom": 16,
            "marginLeft": 16,
          }
        }
      >
        Add
      </Component>
    </SplitItem>
  </Split>
  <div
    style={
      Object {
        "overflowY": "auto",
      }
    }
  >
    <DataList
      aria-label="List of vehicles"
    >
      <Vehicle
        capacity={5}
        capacityChangeHandler={[Function]}
        description="Vehicle 1"
        id={1}
        key="1"
        removeHandler={[MockFunction]}
      />
      <Vehicle
        capacity={5}
        capacityChangeHandler={[Function]}
        description="Vehicle 2"
        id={2}
        key="2"
        removeHandler={[MockFunction]}
      />
    </DataList>
  </div>
</Fragment>
`;
